default/AI/ShipDesignAI.py

```diff
@@ -0,0 +1,1602 @@
+                if new_part.partClass in part_class:
+                    for old_part in [_get_part_type(part) for part in self.strictly_worse_parts
+                                     if part != new_part.name]:
+                        if old_part.partClass in part_class:
+                            if new_part.capacity >= old_part.capacity:
+                                a = new_part
+                                b = old_part
+                            else:
+                                a = old_part
+                                b = new_part
+                            if (self.production_cost[pid][a.name] <= self.production_cost[pid][b.name]
+                                    and {x for x in a.mountableSlotTypes} >= {x for x in b.mountableSlotTypes}):
+                                # ToDo: Maybe add production_time as additional condition?
+                                self.strictly_worse_parts[a.name].append(b.name)
+                    break
+        available_parts = sorted(self.strictly_worse_parts.keys(),
+                                 key=lambda item: _get_part_type(item).capacity, reverse=True)
+
+        # in case of a load, we need to rebuild our Cache.
+        if not self.testhulls:
+            print "Testhull cache not found. This may happen only at first turn after game start or load."
+            for hullname in available_hulls:
+                des = [des for des in testdesign_names_part if des.endswith(hullname)]
+                if des:
+                    self.testhulls.add(hullname)
+            if verbose:
+                print "Rebuilt Cache. The following hulls are used in testdesigns for parts: ", self.testhulls
+
+        # 1. Update hull test designs
+        print "Updating Testdesigns for hulls..."
+        if verbose:
+            print "Available Hulls: ", available_hulls
+            print "Existing Designs (prefix: %s): " % TESTDESIGN_NAME_HULL,
+            print [x.replace(TESTDESIGN_NAME_HULL, "") for x in testdesign_names_hull]
+        for hull in [get_hulltype(hullname) for hullname in available_hulls
+                     if "%s_%s" % (TESTDESIGN_NAME_HULL, hullname) not in testdesign_names_hull]:
+            num_slots = len([slot for slot in hull.slots])
+            partlist = num_slots * [""]
+            testdesign_name = "%s_%s" % (TESTDESIGN_NAME_HULL, hull.name)
+            try:
+                lcopy = list(partlist)
+                res = fo.issueCreateShipDesignOrder(testdesign_name, "TESTPURPOSE ONLY", hull.name,
+                                                    lcopy, "", "fighter", False)
+                del lcopy
+                if res == 1:
+                    print "Success: Added Test Design %s, with result %d" % (testdesign_name, res)
+                else:
+                    print "Failure: When adding test design %s - got result %d but expected 1" % (testdesign_name, res)
+                    continue
+            except Exception:
```

Comment:
the issue* methods try to be careful so as to not generate exceptions, but they don't employ a try:catch wrapper so it is not really guaranteed they wouldn't throw.   A lot of these try:except wrappers are holdovers from previous days when everything was more buggy.   At this point in the game they are probably no more likely to throw and exception than any of the other interface calls that we generally don't wrap with try:except, and I would be fine with not wrapping them.